libantworld/agent.h

```diff
@@ -21,67 +35,279 @@ using namespace units::literals;
+    {
+        BOB_ASSERT(clockwiseSpeed >= -1.f && clockwiseSpeed <= 1.f);
+
+        std::lock_guard<std::mutex> guard(m_PoseMutex);
+        BOB_ASSERT(m_Pose.pitch() == 0_deg && m_Pose.roll() == 0_deg);
+
+        updatePose();
+        m_MoveMode = MoveMode::Turning;
+        m_MoveSpeed = clockwiseSpeed;
+    }
+
+    static auto initialiseWindow(const cv::Size &size)
+    {
+        // Set GLFW error callback
+        glfwSetErrorCallback(handleGLFWError);
+
+        // Initialize the library
+        if (!glfwInit()) {
+            throw std::runtime_error("Failed to initialize GLFW");
+        }
+
+        // Prevent window being resized
+        glfwWindowHint(GLFW_RESIZABLE, false);
+
+        GLFWwindow *ptr = glfwCreateWindow(size.width, size.height, "Ant world", nullptr, nullptr);
+        if (!ptr) {
+            glfwTerminate();
+            throw std::runtime_error("Failed to create window");
+        }
+
+        // Wrap in a unique_ptr so we can free it properly when we're done
+        std::unique_ptr<GLFWwindow, std::function<void(GLFWwindow *)>> window(ptr, &glfwDestroyWindow);
+
+        // Make the window's context current
+        glfwMakeContextCurrent(window.get());
+
+        // Initialize GLEW
+        if (glewInit() != GLEW_OK) {
+            throw std::runtime_error("Failed to initialize GLEW");
+        }
+
+        // Enable VSync
+        glfwSwapInterval(1);
+
+        glDebugMessageCallback(handleGLError, nullptr);
+
+        // Set clear colour to match matlab and enable depth test
+        glClearColor(0.0f, 1.0f, 1.0f, 1.0f);
+        glEnable(GL_DEPTH_TEST);
+        glLineWidth(4.0);
+        glPointSize(4.0);
+
+        glEnable(GL_CULL_FACE);
+        glCullFace(GL_BACK);
+
+        glEnable(GL_TEXTURE_2D);
+
+        return window;
+    }
+
 private:
-    Vector3<degree_t> m_Attitude{{ 0_deg, 0_deg, 0_deg }};
-    Vector3<meter_t> m_Position{{ 0_m, 0_m, 0_m }};
+    using TimeType = std::chrono::time_point<std::chrono::high_resolution_clock>;
+
+    Pose3<meter_t, degree_t> m_Pose;
+    meters_per_second_t m_Velocity;
+    radians_per_second_t m_TurnSpeed;
     Renderer &m_Renderer;
     GLFWwindow *m_Window;
+    std::mutex m_PoseMutex;
+    float m_JoystickX = 0.f, m_JoystickY = 0.f;
+
+    Stopwatch m_MoveStopwatch;
+    enum class MoveMode
+    {
+        NotMoving,
+        MovingForward,
+        Turning
+    } m_MoveMode = MoveMode::NotMoving;
+    float m_MoveSpeed;
+
+    void updatePose()
+    {
+        const units::time::second_t elapsed = m_MoveStopwatch.lap();
+
+        using namespace units::math;
+        switch (m_MoveMode) {
+        case MoveMode::MovingForward: {
+            const meter_t dist = m_MoveSpeed * m_Velocity * elapsed;
+            m_Pose.x() += dist * cos(m_Pose.yaw());
+            m_Pose.y() += dist * sin(m_Pose.yaw());
+        } break;
+        case MoveMode::Turning:
+            m_Pose.yaw() -= m_MoveSpeed * m_TurnSpeed * elapsed;
```

Comment:
How come you're multiplying by both ``m_MoveSpeed`` and ``m_TurnSpeed``?

Reply:
They're just badly named variables. ``m_MoveSpeed`` is a proportion and ``m_TurnSpeed`` is the max turning speed. I'll make it more explicit